File: 0871-minimum-number-of-refueling-stops/0871-minimum-number-of-refueling-stops.java
import java.util.PriorityQueue;
import java.util.Collections;

class Solution {
    public int minRefuelStops(int target, int startFuel, int[][] stations) {
        // Max-heap to store the fuel of stations we have bypassed
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        
        int currentReach = startFuel;
        int stops = 0;
        int index = 0;
        
        // We simulate the journey until our reach is >= the target
        while (currentReach < target) {
            // While there are stations within our current reach, add their fuel to the heap
            while (index < stations.length && stations[index][0] <= currentReach) {
                maxHeap.offer(stations[index][1]);
                index++;
            }
            
            // If we can't reach the target and have no stations to refuel from, it's impossible
            if (maxHeap.isEmpty()) {
                return -1;
            }
            
            // Retrospectively refuel from the station with the most fuel
            currentReach += maxHeap.poll();
            stops++;
        }
        
        return stops;
    }
}

File: 0871-minimum-number-of-refueling-stops/0871-minimum-number-of-refueling-stops.cpp
class Solution {
public:
    int minRefuelStops(int target, int startFuel,
                       vector<vector<int>>& stations) {

        priority_queue<int> maxHeap;

        long long fuel = startFuel;
        int previousPosition = 0;
        int stops = 0;

        for (const auto& station : stations) {

            int position = station[0];
            int stationFuel = station[1];

            fuel -= position - previousPosition;

            while (fuel < 0 && !maxHeap.empty()) {
                fuel += maxHeap.top();
                maxHeap.pop();
                stops++;
            }

            if (fuel < 0) {
                return -1;
            }

            maxHeap.push(stationFuel);
            previousPosition = position;
        }

        fuel -= target - previousPosition;

        while (fuel < 0 && !maxHeap.empty()) {
            fuel += maxHeap.top();
            maxHeap.pop();
            stops++;
        }

        return fuel >= 0 ? stops : -1;
    }
};